### ISSUE TITLE
da.cache

### ISSUE THREAD

**Issue by mrocklin** [ContinuumIO/dask#293]
Should we provide a mechanism to cache dask.arrays?  This would compute and store the array and then create a new Array with a fresh graph.

```python
def cache(self, storage=None):
    """ Return new dask.array with empty graph that points to evaluated array """
    if storage is None:
        storage = ... # some temporary on-disk store
    self.store(storage)
    return da.from_array(storage, chunks=self.chunks)
```

This could be improved by the following:

1.  Figuring out how to do a nice temporary on-disk array store
2.  Supporting multiple inputs/stores at once

**Comment by cowlicks** [comment]
So we would be saving the *graph* generated by daks.array? Not a numpy array?

So as a first approach, saving would  be like this?
```python
def save(self):
    with open(file_name) as f:
        data = pickle.dumps((self.dask, self.keys, self.chunks))
        f.write(data)

**Comment by mrocklin** [comment]
The original idea would be that we would be to trigger computation up to this point, store the result in an array, and then create a new dask.array from that array with a fresh, fairly empty graph.

**Comment by shoyer** [comment]
This is also a great idea. HDF5 is the obvious storage mechanism, but it's out because it only does homogeneously sized chunks.

I would use a hybrid format that stores each chunk in a single binary blob and uses a JSON file to keep track of the index. Optionally, you can compress each chunk (e.g., with gzip or blosc). If you want it to be fully portable, I would store raw bytes (numpy.tostring instead of pickle) in each of the chunk files and put all the shape/dtype metadata in the JSON.

**Comment by mrocklin** [comment]
With the example above we would just use anything that supports the numpy setitem syntax.  HDF5 would do ok, it can handle concurrent writes safely.  We're not being optimally efficient, but that might be ok.

In dask.dataframe.cache we do something more custom like what @shoyer suggests, we dump the values straight into a `chest` without storing metadata in the temporary store.  We copy the metadata directly over to the new dask collection object without serializing it anywhere.

**Comment by shoyer** [comment]
Yeah, actually using chest would be a nice simple way to handle this.

**Comment by thpfaff** [comment]
Couldn't this also be a way to implement the eager evaluation mode of issue #292 ?

**Comment by mrocklin** [comment]
> Couldn't this also be a way to implement the eager evaluation mode of issue #292 ?

Yes.  That sounds like a great idea.

**Comment by mrocklin** [comment]
This is possibly resolved in #299
